e2e-test/log/src/integrationTest/java/com/amazon/dataprepper/integration/log/EndToEndBasicGrokTest.java

```diff
@@ -0,0 +1,139 @@
+import com.linecorp.armeria.common.RequestHeaders;
+import com.linecorp.armeria.common.SessionProtocol;
+import org.junit.Assert;
+import org.junit.Test;
+import org.opensearch.action.admin.indices.refresh.RefreshRequest;
+import org.opensearch.action.search.SearchRequest;
+import org.opensearch.action.search.SearchResponse;
+import org.opensearch.client.RequestOptions;
+import org.opensearch.client.RestHighLevelClient;
+import org.opensearch.search.SearchHits;
+import org.opensearch.search.builder.SearchSourceBuilder;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
+import static org.awaitility.Awaitility.await;
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.MatcherAssert.assertThat;
+
+public class EndToEndBasicGrokTest {
+    private static final int HTTP_SOURCE_PORT = 2021;
+    private static final String TEST_INDEX_NAME = "test-grok-index";
+
+    private final ApacheLogFaker apacheLogFaker = new ApacheLogFaker();
+    private final ObjectMapper objectMapper = new ObjectMapper();
+
+    @Test
+    public void testPipelineEndToEnd() throws JsonProcessingException {
+        // Send data to http source
+        sendHttpRequestToSource(HTTP_SOURCE_PORT, generateRandomApacheLogHttpData(2));
+        sendHttpRequestToSource(HTTP_SOURCE_PORT, generateRandomApacheLogHttpData(3));
+
+        // Verify data in OpenSearch backend
+        final RestHighLevelClient restHighLevelClient = prepareOpenSearchRestHighLevelClient();
+        final List<Map<String, Object>> retrievedDocs = new ArrayList<>();
+        // Wait for data to flow through pipeline and be indexed by ES
+        await().atMost(10, TimeUnit.SECONDS).untilAsserted(
+                () -> {
+                    refreshIndices(restHighLevelClient);
+                    final SearchRequest searchRequest = new SearchRequest(TEST_INDEX_NAME);
+                    searchRequest.source(
+                            SearchSourceBuilder.searchSource().size(100)
+                    );
+                    final SearchResponse searchResponse = restHighLevelClient.search(searchRequest, RequestOptions.DEFAULT);
+                    final List<Map<String, Object>> foundSources = getSourcesFromSearchHits(searchResponse.getHits());
+                    Assert.assertEquals(5, foundSources.size());
+                    retrievedDocs.addAll(foundSources);
+                }
+        );
+        // Verify original and grokked keys from retrieved docs
+        final List<String> expectedDocKeys = Arrays.asList(
+                "date", "log", "clientip", "ident", "auth", "timestamp", "verb", "request", "httpversion", "response", "bytes");
+        retrievedDocs.forEach(expectedDoc -> {
+            for (String key: expectedDocKeys) {
+                Assert.assertNotNull(expectedDoc.get(key));
```

Comment:
Is there a way to do a check for the values themselves instead of just checking for non-null? Since ordering in sending and retrieving isn't guaranteed I can see this being slightly annoying.

Reply:
I'd recommend:
```
assertThat(expectedDoc, hasKey(key));
assertThat(expectedDoc.get(key), equalTo(PLACEHOLDER));
```

Using `hasKey` is much clearer in conveying intent.

Reply:
@dlvenable @graytaylor0 Thanks for the comments. 

I think hasKey makes a lot of sense. For checking values themselves, it seems adding little value to me b/c essentially it is testing the business logic of grok-prepper(java-grok) which should be covered by the unit test itself, here the main intention is to make sure the logs has been grokked before exporting to opensearch.